File: nhglib-core/src/io/github/movementspeed/nhglib/core/ecs/utils/UiManager.java
package io.github.movementspeed.nhglib.core.ecs.utils;

import com.badlogic.gdx.graphics.Pixmap;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.graphics.glutils.FrameBuffer;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.ArrayMap;
import com.badlogic.gdx.utils.viewport.ExtendViewport;
import io.github.movementspeed.nhglib.assets.Asset;
import io.github.movementspeed.nhglib.input.models.base.NhgInput;
import net.peakgames.libgdx.stagebuilder.core.assets.Assets;
import net.peakgames.libgdx.stagebuilder.core.assets.ResolutionHelper;
import net.peakgames.libgdx.stagebuilder.core.assets.StageBuilderFileHandleResolver;
import net.peakgames.libgdx.stagebuilder.core.builder.StageBuilder;
import net.peakgames.libgdx.stagebuilder.core.services.LocalizationService;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class UiManager {
    private boolean changesOrientation;
    private boolean initiallyEmptyStage;

    private int width;
    private int height;
    private int virtualWidth;
    private int virtualHeight;

    private Stage stage;
    private String fileName;
    private Assets assets;
    private StageBuilderFileHandleResolver fileHandleResolver;
    private StageBuilder stageBuilder;
    private ResolutionHelper resolutionHelper;
    private NhgLocalizationService localizationService;
    private FrameBuffer frameBuffer;
    private TextureRegion textureRegion;

    private Array<String> actorNames;
    private List<Vector2> supportedResolutions;
    private ArrayMap<String, NhgInput> actorInputs;

    public UiManager(String fileName, List<Vector2> supportedResolutions) {
        this(fileName, false, supportedResolutions);
    }

    public UiManager(String fileName, boolean changesOrientation, List<Vector2> supportedResolutions) {
        this(fileName, changesOrientation, false, supportedResolutions);
    }

    public UiManager(String fileName, boolean changesOrientation, boolean initiallyEmptyStage, List<Vector2> supportedResolutions) {
        this.fileName = fileName;
        this.changesOrientation = changesOrientation;
        this.initiallyEmptyStage = initiallyEmptyStage;
        this.supportedResolutions = supportedResolutions;

        Collections.sort(this.supportedResolutions, new Comparator<Vector2>() {
            @Override
            public int compare(Vector2 o1, Vector2 o2) {
                return (int) (o1.x - o2.x);
            }
        });
    }

    public void init(float virtualWidth, float virtualHeight, float width, float height, Array<Asset> dependencies) {
        this.width = (int) width;
        this.height = (int) height;
        this.virtualWidth = (int) virtualWidth;
        this.virtualHeight = (int) virtualHeight;

        this.resolutionHelper = new ResolutionHelper(
                virtualWidth, virtualHeight,
                width, height,
                findBestResolution().x);

        this.localizationService = new NhgLocalizationService();

        fileHandleResolver = new StageBuilderFileHandleResolver(this.width, supportedResolutions);
        this.assets = new Assets(fileHandleResolver, resolutionHelper);

        for (Asset dependency : dependencies) {
            assets.addAssetConfiguration(fileName, dependency.source, dependency.assetClass);
        }

        assets.loadAssetsSync(fileName);

        this.stageBuilder = new StageBuilder(assets, resolutionHelper, localizationService);
        createStage(initiallyEmptyStage);
    }

    public void renderUi(float delta) {
        stage.act(delta);
        stage.draw();
    }

    public void resize(int newWidth, int newHeight) {
        if (this.width == newWidth && this.height == newHeight) {
            return;
        }

        this.width = newWidth;
        this.height = newHeight;

        int newTargetWidth = virtualWidth;
        int newTargetHeight = virtualHeight;

        fileHandleResolver = new StageBuilderFileHandleResolver(this.width, supportedResolutions);

        if (this.height > this.width) {
            newTargetWidth = virtualHeight;
            newTargetHeight = virtualWidth;
        }

        this.resolutionHelper.resize(newTargetWidth,
                newTargetHeight,
                this.width,
                this.height);

        createStage(initiallyEmptyStage);
    }

    public void dispose() {
        stage.dispose();
    }

    /**
     * @return the resolution which has the the closest width value.
     */
    public Vector2 findBestResolution() {
        int minDiff = Integer.MAX_VALUE;
        int bestResIndex = 0;

        for (int i = 0; i < supportedResolutions.size(); i++) {
            int diff = Math.abs(width - (int) supportedResolutions.get(i).x);
            if (diff < minDiff) {
                minDiff = diff;
                bestResIndex = i;
            }
        }

        return supportedResolutions.get(bestResIndex);
    }

    public TextureRegion renderUiToTexture(float delta) {
        if (frameBuffer == null) {
            frameBuffer = new FrameBuffer(Pixmap.Format.RGBA8888, width, height, false);
            textureRegion = new TextureRegion(frameBuffer.getColorBufferTexture());
        }

        frameBuffer.begin();
        renderUi(delta);
        frameBuffer.end();

        textureRegion.setRegion(frameBuffer.getColorBufferTexture());
        //textureRegion.flip(true, true);
        return textureRegion;
    }

    public Stage getStage() {
        return stage;
    }

    private void createStage(final boolean initiallyEmptyStage) {
        if (initiallyEmptyStage) {
            stage = new Stage();
            stage.addActor(stageBuilder.createRootGroup(null));
        } else {
            stage = stageBuilder.build(fileName, new ExtendViewport(width, height));
        }

        stage.setDebugAll(true);
    }

    public class NhgLocalizationService implements LocalizationService {
        @Override
        public String getString(String s) {
            return null;
        }

        @Override
        public String getString(String s, Object... args) {
            return null;
        }
    }
}

File: nhglib-core/src/io/github/movementspeed/nhglib/core/ecs/utils/UiManager.kt
package io.github.movementspeed.nhglib.core.ecs.utils

import com.badlogic.gdx.graphics.Pixmap
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.graphics.glutils.FrameBuffer
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.ArrayMap
import com.badlogic.gdx.utils.viewport.ExtendViewport
import io.github.movementspeed.nhglib.assets.Asset
import io.github.movementspeed.nhglib.input.models.base.NhgInput
import net.peakgames.libgdx.stagebuilder.core.assets.Assets
import net.peakgames.libgdx.stagebuilder.core.assets.ResolutionHelper
import net.peakgames.libgdx.stagebuilder.core.assets.StageBuilderFileHandleResolver
import net.peakgames.libgdx.stagebuilder.core.builder.StageBuilder
import net.peakgames.libgdx.stagebuilder.core.services.LocalizationService

import java.util.Collections
import java.util.Comparator

class UiManager(private val fileName: String, private val changesOrientation: Boolean, private val initiallyEmptyStage: Boolean, private val supportedResolutions: List<Vector2>) {

    private var width: Int = 0
    private var height: Int = 0
    private var virtualWidth: Int = 0
    private var virtualHeight: Int = 0

    var stage: Stage? = null
        private set
    private var assets: Assets? = null
    private var fileHandleResolver: StageBuilderFileHandleResolver? = null
    private var stageBuilder: StageBuilder? = null
    private var resolutionHelper: ResolutionHelper? = null
    private var localizationService: NhgLocalizationService? = null
    private var frameBuffer: FrameBuffer? = null
    private var textureRegion: TextureRegion? = null

    private val actorNames: Array<String>? = null
    private val actorInputs: ArrayMap<String, NhgInput>? = null

    constructor(fileName: String, supportedResolutions: List<Vector2>) : this(fileName, false, supportedResolutions) {}

    constructor(fileName: String, changesOrientation: Boolean, supportedResolutions: List<Vector2>) : this(fileName, changesOrientation, false, supportedResolutions) {}

    init {

        Collections.sort(this.supportedResolutions) { o1, o2 -> (o1.x - o2.x).toInt() }
    }

    fun init(virtualWidth: Float, virtualHeight: Float, width: Float, height: Float, dependencies: Array<Asset>) {
        this.width = width.toInt()
        this.height = height.toInt()
        this.virtualWidth = virtualWidth.toInt()
        this.virtualHeight = virtualHeight.toInt()

        this.resolutionHelper = ResolutionHelper(
                virtualWidth, virtualHeight,
                width, height,
                findBestResolution().x)

        this.localizationService = NhgLocalizationService()

        fileHandleResolver = StageBuilderFileHandleResolver(this.width, supportedResolutions)
        this.assets = Assets(fileHandleResolver, resolutionHelper)

        for (dependency in dependencies) {
            assets!!.addAssetConfiguration(fileName, dependency.source, dependency.assetClass)
        }

        assets!!.loadAssetsSync(fileName)

        this.stageBuilder = StageBuilder(assets, resolutionHelper, localizationService)
        createStage(initiallyEmptyStage)
    }

    fun renderUi(delta: Float) {
        stage!!.act(delta)
        stage!!.draw()
    }

    fun resize(newWidth: Int, newHeight: Int) {
        if (this.width == newWidth && this.height == newHeight) {
            return
        }

        this.width = newWidth
        this.height = newHeight

        var newTargetWidth = virtualWidth
        var newTargetHeight = virtualHeight

        fileHandleResolver = StageBuilderFileHandleResolver(this.width, supportedResolutions)

        if (this.height > this.width) {
            newTargetWidth = virtualHeight
            newTargetHeight = virtualWidth
        }

        this.resolutionHelper!!.resize(newTargetWidth.toFloat(),
                newTargetHeight.toFloat(),
                this.width.toFloat(),
                this.height.toFloat())

        createStage(initiallyEmptyStage)
    }

    fun dispose() {
        stage!!.dispose()
    }

    /**
     * @return the resolution which has the the closest width value.
     */
    fun findBestResolution(): Vector2 {
        var minDiff = Integer.MAX_VALUE
        var bestResIndex = 0

        for (i in supportedResolutions.indices) {
            val diff = Math.abs(width - supportedResolutions[i].x.toInt())
            if (diff < minDiff) {
                minDiff = diff
                bestResIndex = i
            }
        }

        return supportedResolutions[bestResIndex]
    }

    fun renderUiToTexture(delta: Float): TextureRegion {
        if (frameBuffer == null) {
            frameBuffer = FrameBuffer(Pixmap.Format.RGBA8888, width, height, false)
            textureRegion = TextureRegion(frameBuffer!!.colorBufferTexture)
        }

        frameBuffer!!.begin()
        renderUi(delta)
        frameBuffer!!.end()

        textureRegion!!.setRegion(frameBuffer!!.colorBufferTexture)
        //textureRegion.flip(true, true);
        return textureRegion
    }

    private fun createStage(initiallyEmptyStage: Boolean) {
        if (initiallyEmptyStage) {
            stage = Stage()
            stage!!.addActor(stageBuilder!!.createRootGroup(null))
        } else {
            stage = stageBuilder!!.build(fileName, ExtendViewport(width.toFloat(), height.toFloat()))
        }

        stage!!.isDebugAll = true
    }

    inner class NhgLocalizationService : LocalizationService {
        override fun getString(s: String): String? {
            return null
        }

        override fun getString(s: String, vararg args: Any): String? {
            return null
        }
    }
}
